src/Simulation/Microsoft.Quantum.TargetGeneration/GenerateTarget.fs

```diff
@@ -0,0 +1,398 @@
+                                      | QsAdjoint           -> "Adj"
+                                      | QsControlled        -> "Ctl"
+                                      | QsControlledAdjoint -> "CtlAdj"
+
+    let fullArgsForSpecialization args (sp : QsSpecialization) =
+        match sp.Kind with
+        | QsControlled | QsControlledAdjoint -> ("__ctrls__", "IQArray<Qubit>") :: args
+        | _ -> args
+
+    let buildImplMethod context args (sp : QsSpecialization) =
+        let outType = match sp.Signature.ReturnType.Resolution with
+                      | UnitType -> "void"
+                      | _ -> SimulationCode.roslynTypeName context sp.Signature.ReturnType
+        match sp.Implementation with
+        | Generated SelfInverse -> Seq.empty
+        | _ ->
+            let implName = implMethodName sp
+            let fullArgs = fullArgsForSpecialization args sp
+            let parms = fullArgs |> List.map (fun (n, t) -> ``param`` n ``of`` (``type`` t))
+            let impl = ``method`` outType implName ``<<`` [] ``>>`` 
+                            ``(`` parms ``)`` [``public``;``abstract``]
+                            ``{`` [] ``}`` :> MemberDeclarationSyntax
+            seq { impl }
+
+    let buildImplProperty targetClass context args (sp : QsSpecialization) =
+        let inType  = SimulationCode.roslynTypeName context sp.Signature.ArgumentType
+        let outType = SimulationCode.roslynTypeName context sp.Signature.ReturnType
+        let propertyType = "Func<" + inType + ", " + outType + ">"
+        let propertyName =
+            match sp.Kind with
+            | QsBody              -> "Body"
+            | QsAdjoint           -> "AdjointBody"
+            | QsControlled        -> "ControlledBody"
+            | QsControlledAdjoint -> "ControlledAdjointBody"
+        let propertyFromBody body =
+            (``property-arrow_get`` propertyType propertyName [``public``; ``override``]
+                    ``get`` (``=>`` body)) :> MemberDeclarationSyntax
+        match sp.Implementation with
+        | SpecializationImplementation.Intrinsic -> 
+            let returnType  = sp.Signature.ReturnType
+            let inData = ``ident`` "__in__"
+            let spArgs = match sp.Kind with 
+                         | QsControlled | QsControlledAdjoint -> 
+                            let controls = 
+                                { 
+                                    VariableName = NonNullable<string>.New "__ctrls__" |> ValidName
+                                    Type = ResolvedType.New (ArrayType (ResolvedType.New Qubit))
+                                    InferredInformation = InferredExpressionInformation.New (false, false)
+                                    Position = Null
+                                    Range = (QsPositionInfo.Zero, QsPositionInfo.Zero)
+                                }
+                                |> QsTupleItem
+                            match args with
```

Comment:
I am not following this: Why is this a flattened argument list? 
And as an edge case: for an operation Op taking unit as argument, Controlled Op would actually require a tuple of the control qubits and a unit value.

Reply:
The flattening  is copied from the C# code gen. 
Controlled Op(), when Op takes Unit, is a bug; I'll fix it and add a unit test. It's a little odd -- a quantum-controlled random number generator, perhaps, but that's weird -- but it should still be processed correctly.

Reply:
OK, Controlled Op when Op takes no parameters is now working, I think.

Reply:
Thanks!